Sources/SPMLLBuild/llbuild.swift

```diff
@@ -8,10 +8,138 @@
  See http://swift.org/CONTRIBUTORS.txt for Swift project authors
 */
 
+import class Foundation.JSONEncoder
+import class Foundation.JSONDecoder
+import struct Foundation.Data
+
 #if canImport(llbuildSwift)
 @_exported import llbuildSwift
 #elseif canImport(llbuild)
 @_exported import llbuild
 #else
 // This should be a hard error in future.
 #endif
+
+public protocol LLBuildKey: Codable {
+}
+
+public protocol LLBuildValue: Codable {
+}
+
+public protocol LLBuildEngineDelegate {
+    func lookupRule(rule: String, key: Key) -> Rule
+}
+
+public final class LLBuildEngine {
+
+    private final class Delegate: BuildEngineDelegate {
+        let delegate: LLBuildEngineDelegate
+
+        init(_ delegate: LLBuildEngineDelegate) {
+            self.delegate = delegate
+        }
+
+        func lookupRule(_ key: Key) -> Rule {
+            let ruleKey = RuleKey(key)
+            return delegate.lookupRule(
+                rule: ruleKey.rule, key: Key(ruleKey.data))
+        }
+    }
+
+    private let engine: BuildEngine
+
+    public init(delegate: LLBuildEngineDelegate) {
+        engine = BuildEngine(delegate: Delegate(delegate))
+    }
+
+    public func build(key: Key) -> Value {
```

Comment:
I feel like this can be more generic...

It feels to me like everything should start with a generic Key protocol. The Key protocol should have an associated type for the kind of Value it produces. It should also have a field for the Rule instance to compute it (a static field pointing to the type maybe).

There are some details to work out but I think this could eliminate the manual coding that shows up the implementation. Just a thought, it’s probably overkill for now but could be nice as llbuild clients on SwiftPM grow.